app_legacy/Helpers/render/code-note.php

```diff
@@ -1,41 +1,63 @@
 <?php
 
-function RenderCodeNotes(array $codeNotes): void
+function RenderCodeNotes(array $codeNotes, bool $editable = false): void
 {
-    echo "<table class='table-highlight'><tbody>";
+    echo "<table class='table-highlight'>";
 
-    echo "<tr class='do-not-highlight'><th style='font-size:100%;'>Mem</th><th style='font-size:100%;'>Note</th><th style='font-size:100%;'>Author</th></tr>";
+    echo "<thead>";
+    echo "<tr class='do-not-highlight'>";
+    echo "<th style='font-size:100%;'>Mem</th>";
+    echo "<th style='font-size:100%;'>Note</th>";
+    echo "<th style='font-size:100%;'>Author</th>";
+    if ($editable) {
+        echo "<th>Dev</th>";
+    }
+    echo "</tr>";
+    echo "</thead>";
+
+    echo "<tbody>";
 
+    $rowIndex = 0;
     foreach ($codeNotes as $nextCodeNote) {
         if (empty(trim($nextCodeNote['Note'])) || $nextCodeNote['Note'] == "''") {
             continue;
         }
 
-        echo "<tr>";
+        echo "<tr id='row-$rowIndex'>";
 
         $addr = $nextCodeNote['Address'];
         $addrInt = hexdec($addr);
 
         $addrFormatted = sprintf("%04x", $addrInt);
-        $memNote = $nextCodeNote['Note'];
+        $originalMemNote = $nextCodeNote['Note'];
 
-        sanitize_outputs($memNote);
+        sanitize_outputs($originalMemNote);
 
-        $memNote = nl2br($memNote);
+        $memNote = nl2br($originalMemNote);
 
-        echo "<td style='width: 25%;'>";
+        echo "<td data-address='$addr' style='width: 25%;'>";
         echo "<span class='font-mono'>0x$addrFormatted</span>";
         echo "</td>";
 
         echo "<td>";
-        echo "<div class='font-mono' style='word-break:break-word'>$memNote</div>";
+        echo "<div class='font-mono note-display block' style='word-break:break-word'>$memNote</div>";
+        echo "<textarea class='w-full font-mono note-edit hidden'>$originalMemNote</textarea>";
         echo "</td>";
 
         echo "<td>";
         echo userAvatar($nextCodeNote['User'], label: false, iconSize: 24);
         echo "</td>";
 
+        if ($editable) {
+            echo "<td>";
+            echo "<button class='edit-btn inline' onclick='beginEditMode($rowIndex)'>Edit</button>";
+            echo "<button class='save-btn hidden' onclick='saveCodeNote($rowIndex)'>Save</button>";
```

Comment:
There is no clean way to cancel an edit. Saving without changes still transfers ownership to the current user.

Reply:
Both of these pieces of feedback have now been addressed.

A cancel button has been added. The save button has also been relocated.

<img width="856" alt="Screenshot 2023-03-20 at 9 33 44 PM" src="https://user-images.githubusercontent.com/3984985/226499596-805053f8-6d2d-454b-bfe9-946dc49797e3.png">

I opted to put the Save button underneath the textarea to prevent accidental clicks. Cancel is now where Save previously was.

If the user saves without changes, it is treated identically to a Cancel button click.

Reply:
The Save button slightly overlaps the field when focused
![image](https://user-images.githubusercontent.com/32680403/226970440-86ac0188-b8ee-48be-ab7e-4b41b09c4587.png)

Reply:
This should be addressed in https://github.com/RetroAchievements/RAWeb/pull/1440/commits/3b017ef1ac1b059a4ccb62a584bf8a93ced8e4b1. I've added 6px of margin between the button and the input field.